### ISSUE TITLE
constexpr functions

### ISSUE THREAD

**Issue by MikeGitb** [isocpp/CppCoreGuidelines#964]
First of all there is a bug in [F.8](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f8-prefer-pure-functions), where it states 

> `constexpr` functions are **pure**.

Counterexample:

    int param = 0;
    constexpr int foo(bool arg) {
        if (arg) {
            return 0;
        } else {
            return param*param;
        }
    }

    int bar() {
        param = 2;
        // result depends on global variable 2
        return foo(false); 
    }

I have the general impression that there is a lot of confusion in the c++ community at large, but even in the core guidelines itself, about what the properties of constexpr functions are.

Common misconceptions I've seen:
1. constexpr functions are always evaluated during compiletime
2. whether or not a constexpr function is evaluated during compiletime depends **ONLY** on whether the arguments are compiletime constants or not (imho also implied in [F.8](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f8-prefer-pure-functions))
2. constexpr functions can't throw (see https://github.com/isocpp/CppCoreGuidelines/issues/616)
3. constexpr functions are pure (as just mentioned above)

3 and 4 are probably a result of 1.

At the very least, the guidelines need to be fixed (I'm not sure if there are any more mistakes but at a glance I didn't find any). 
Ideally, there should be some guideline or a set of guidelines that summarizes the properties of constexpr functions, but I'm not sure if/how this fits best into the format of the cppcoreguidelines.

A final question for those that have experience with non-trivial production application: How often can you actually use the constexpr-ness of a function outside of template meta programming? I got the impression that almost all data comes from some runtime source (user input, configuration files, networks, database etc.) and hence, very few computations can actually be performed at compile time.

**Comment by jwakely** [comment]
> Ideally, there should be some guideline or a set of guidelines that summarizes the properties of constexpr functions, but I'm not sure if/how this fits best into the format of the cppcoreguidelines.

That seems more appropriate for https://isocpp.org/wiki/faq/

**Comment by jwakely** [comment]
> How often can you actually use the constexpr-ness of a function outside of template meta programming?

Plenty of other places. Initializing static data before execution starts (replacing traditional uses of macros). Calculating constants such as the number of elements to define an array with, if it depends on compile-time values such as sizeof values that can vary between platforms. Many of these can be done with macros or arbitrarily complex expressions, but constexpr functions allow the calculation to be moved out into a function, allowing more structured programming.

**Comment by AndrewPardoe** [comment]
Per our editor's discussion, there is a mistake in the doc: `constexpr` functions can depend upon global state. @gdr-at-ms, do you have comments on the rest of this?

**Comment by AndrewPardoe** [comment]
Per our editor's discussion, `constexpr` functions are not pure, but we do want to say something. @MikeGitb, we're going to reject your PR because while the text is wrong it is a valuable note. @gdr-at-ms, can you please create a new PR for this?

**Comment by MikeGitb** [comment]
Sure, whatever you feel is best. 
I should probably mention that I only checked if above code compiles and behaves as I expected, but I did not actually check the standard whether it is legal.

> while the text is wrong it is a valuable note.

Isn't that contradictory?

**Comment by jwakely** [comment]
It's not all wrong, parts of it (and the overall point) are valuable.

**Comment by MikeGitb** [comment]
Sorry, I'm lost. What do you consider the overall point?

Those are the lines I proposed to delete:

> `constexpr` functions are pure.

Afaik that is not correct and the main point of this issue

> When given a non-constant argument, a `constexpr` function can throw.

It is correct that a constexpr function can throw but I don't see how that is important here. Also, the sentence seems to imply that a `constexpr` functions can not throw if it is given a constant argument. 
Correct me if I'm wrong, but IIRC,, constexpr arguments don't guarantee execution at compiletime (which is imho very unfortunate) so this is also not true.

> If you consider exiting by throwing a side effect, a `constexpr` function isn't completely pure;
> if not, this is not an issue.

I actually consider that (whether a throwing function is considered pure) an interesting question, but it is a question not a guideline - so what is its point here?

> ??? A question for the committee: can a constructor for an exception thrown by a `constexpr` function modify state?
-"No" would be a nice answer that matches most practice.

Again, an interesting question ( I wouldn't know why it can't but I'm not a language laywer) but no guideline.

**Comment by MikeGitb** [comment]
Btw.: Maybe a good guideline would be something along the lines of

> Avoid writing constexpr functions that are impure when executed during runtime, because users generally don't expect them to behave differently during runtime.

**Comment by jwakely** [comment]
The overall point is to prefer pure functions. Discussing constexpr functions in that context is useful (even though not all constexpr functions are pure, as it incorrectly states at the moment).

> Correct me if I'm wrong, but IIRC,, constexpr arguments don't guarantee execution at compiletime (which is imho very unfortunate) so this is also not true.

Explaining the semantics of constexpr is off-topic here, but initialization of constexpr variables is guaranteed to happen before execution. Other constexpr evaluations _can_ be done during translation, which is good enough in practice.

> As I said, do whatever you think best - I'm just looking for clarification.

Maybe a little patience? It's been assigned to somebody to provide new wording, so you'll get it then.

**Comment by MikeGitb** [comment]
> Maybe a little patience? It's been assigned to somebody to provide new wording, so you'll get it then.

Well, it has been assigned to someone 3 months ago and the last time I reported a similar issue it took almost a year until the text was changed. 
I know you guys are busy and I thought my chances of understanding the problem with my PR in a timely manner are higher, if I'm asking here instead of waiting on a future change that may or may not explain it.

However, I got the message - I'll just wait and see.

**Comment by MikeGitb** [comment]
Ping on this - I remember a couple of posts at Reddit 2-3 months ago, that precisely demonstrated my concern, namely that a lot of c++ programmers interpret more into the constexpr keyword than what they actually guarantees. The coreguidelines  spreading misinformation doesn't really help (although I doubt they are a significant factor in this regard either).

**Comment by AndrewPardoe** [comment]
Thanks for the ping, @MikeGitb. I remember this discussion: we really wanted Gaby to own this issue. He's been a bit busy with C++ Modules over the past few months but I will ping him about it.

**Comment by MikeGitb** [comment]
Thats ok. As far as I am concerned, modules trump (almost) anything else priority wise.

**Comment by MikeGitb** [comment]
My regular ping on this.

What I really don't understand: Removing those wrong/irrelevant sentences would make the guidelines strictly better (as I have explained line by line), would come at almost no effort (I know you have more important things to do) and doesn't prevent a "better" writeup in the future if so desired and someone finally comes around to do it. Still, you keep insisting on keeping this erroneous information around for almost a year in the hope that someone might find the time to replace it with something better. 

Is this a case where the better is the enemy of the good?

**Comment by MikeGitb** [comment]
Btw., that mistake is repeated in [F.4](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f4-if-a-function-may-have-to-be-evaluated-at-compile-time-declare-it-constexpr)

**Comment by MikeGitb** [comment]
Closing due to inactivity. If someone else wants to track this, feel free to open a different issue.
